doc/tinyproxy.8
.\" -*- nroff -*-
.\"
.\" tinyproxy.1
.\"
.\" Copyright (c) 1999, Robert James Kaes, Steven Young.
.\"
.\" This program is distributed under the terms of the GNU General Public
.\" License. See COPYING for additional information.
.\"
.TH tinyproxy 1 "December 23, 1999" "tinyproxy" "tinyproxy Manual"
.SH NAME
tinyproxy \- a small HTTP proxy server
.SH SYNOPSIS
tinyproxy [options]
.SH DESCRIPTION
.B tinyproxy
is an HTTP proxy server designed to consume a minimum of system resources.
It listens on a given TCP port and handles HTTP proxy requests.
.SH OPTIONS
.TP 5
.BI "\-a headername"
Allow the header
.I headername
through to the server (anon mode only).  It is highly recommended that you
specify -a Host: and -a Authorization:, otherwise HTTP/1.1 virtual domains
and basic authentication will fail. If anon mode is compiled into the
program, but you do not provide a
.I headername
tinyproxy will run as if anon mode was not compiled in.
.TP 5
.BI "\-d" 
Don't daemonize; stay in the foreground.
.TP 5
.BI "\-f filterfile"
Block any sites/URL which are specified in the
.I filterfile
The domains/URLs are specified in a flat text file with one filter per line.
.TP 5
.BI "\-h"
Display a short help screen of command line arguments and exit.
.TP 5
.BI "\-i ip_address"
Only listen to the specific address. Useful for systems with more than one
network connection.
.I ip_address
needs to be in the dotted-decimal form (e.g. 192.168.0.1)
.TP 5
.BI "\-l logfile"
Log to the file
.I logfile
instead of the default
.B /var/log/tinyproxy.
The default can be changed at compile time using the
.B "\-\-with\-log\-file"
directive.
.TP 5
.BI "\-n ip_address"
Allow access from only those hosts in the subnet.
.I ip_address
should be of a form xxx.xxx.xxx. like 192.168.0.
.TP 5
.BI "\-p port"
Listen to
.I port
instead of the default
.B 8888.
The default can be changed at compile time by using the
.B "\-with\-port"
directive.
.TP 5
.BI "\-r"
Restrict the log file to only logging error conditions. The hosts connecting
and the locations being accessed are not logged.
.TP 5
.BI "\-s stathost"
Set the status host to
.I stathost
.TP 5
.BI "\-S"
Log to the syslog daemon
.TP 5
.BI "\-t domain:port"
If you would like to use an upstream proxy along with tinyproxy then set
this option. The
.I domain
is the address of the proxy, and the
.I port
is the
decimal port number. The address can be either alphanumeric, or dotted-decimal.
.TP 5
.BI "\-u username"
Begin running as
.I username
after the port has been bound and the logfile has been opened.  Only
works if tinyproxy is being run as root.  "" to disable.
.TP 5
.BI "\-v"
Display version information and exit.
.TP 5
.BI "\-w loadaverage"
Stop handling requests when the load is
.I loadaverage
or higher.  Specifying a
.I loadaverage
of 0.00 will disable this feature.
(disabled by default)
.TP 5
.BI "\-x domain"
Add a
.B XTinyproxy
header with the peer's IP address.
.I domain
tells tinyproxy which domains to include the
.B XTinyproxy
header on. Basically, set this if you want to include the header only on
your sites. Else, set to "" to have it sent on all connections.
.SH SIGNALS
In addition to these command line options, there are also several signals
that can be sent to tinyproxy while it is running to generate debugging
information and to force certain events.
.TP 5
.B SIGUSR1
Dump a summary of all current connections to the logfile.
.TP 5
.B SIGHUP
Force
.B tinyproxy
to do a garbage collection on the current connections linked list.  This
is usually done automatically after a certain number of connections have
been handled.
.SH FILES
.SM /var/log/tinyproxy.log
.SH AUTHORS
Steven Young (sdyoung@well.com)
.PP
Robert James Kaes (rjkaes@flarenet.com)
.SH BUGS
For some reason, when sent SIGTERM,
.B tinyproxy 
occasionally waits for the
next incoming connection before exiting.  We believe this bug has 
been corrected, but please e-mail the authors if you see this 
behaviour occurring.
.SH COPYRIGHT
.B tinyproxy
is distributed under the GNU Public License (GPL).  For more information
on the GPL, please see the file COPYING which should have been included in
the archive with tinyproxy.  Failing that, http://www.fsf.org/ will
doubtless have a copy up for you to peruse.  Please don't use this
software if you don't agree to the terms specified therein.
.SH AVAILABILITY
The latest version of
.B tinyproxy
can be acquired from:
.PP
http://www.flarenet.com/tinyproxy/
http://tinyproxy.attacca.com/


NEWS
Tinyproxy NEWS
==============

Version 1.8.2
-------------

 * Minor formatting changes and typo fixes were made.

Bugs fixed since version 1.8.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 * BB#69: INET6 not available when configured to Listen and Bind in v4,
   and vice versa
 * BB#74: tinyproxy unable to reopen log files after receiving HUP
 * BB#78: Warn if configuration results in an open proxy
 * BB#82: https access not working
 * BB#83: run_tests.sh relies on $USER
 * BB#84: Unaligned access error on ia64 and alpha
 * BB#87: Unable to listen on ports less than 1024 (regression in 1.8.1)
 * BB#88: Crashes when reloading configuration
 * BB#89: tinyproxy leaks memory over time

Contributors
~~~~~~~~~~~~

Dmitry Semyonov, John van der Kamp, Jordi Mallach, Michael Adam,
Mukund Sivaraman.


Version 1.8.1
-------------

 * Tinyproxy now drops `root` user privileges more quickly.
 * The log and pid files are now stored in a sub-directory in `/var/`.
 * A format string vulnerability was fixed.
 * Minor formatting changes and typo fixes were made.

Bugs fixed since version 1.8.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 * BB#74: tinyproxy unable to reopen log files after receiving HUP
 * BB#79: Make the testsuite uninteractive
 * BB#80: Handle errors in testsuite
 * BB#81: Listen directive doesn't work as expected
 * BB#72: upstream support is not reported with tinyproxy -h
 * BB#73: generated tinyproxy.conf has the wrong location for the html
   file installation

Contributors
~~~~~~~~~~~~

Michael Adam, Mukund Sivaraman.


Version 1.8.0
-------------

 * Tinyproxy now reloads its configuration upon SIGHUP signal.
 * Tinyproxy reopens its log file (instead of truncation) upon SIGHUP
   signal. This is to play more nicely with logrotate.
 * File logging is now the default.
   Syslog is chosen if and only if "SysLog Yes" is in the config,
   i.e., a present "SysLog Yes" in the config file now overrides
   any LogFile setting.
 * The XTinyProxy option is now documented as a global boolean.
   Before it was documented to build a list of sites to add a
   X-Tinyproxy header for, but it was implemented as global boolean.
 * A new config option AddHeader allows the user to configure a list of
   custom headers to send in outgoing HTTP requests.
 * A new config option DisableViaHeader allows the user to disable
   sending of the "Via:" header.
 * Tinyproxy is now IPv6 capable.
 * The config option PidFile now has a compiled in default.

Bugs fixed since version 1.7.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 * BB#9: Add support for the IPv6 protocol
 * BB#17: Add support for custom headers
 * BB#55: Error message response omits body when request has a body
 * BB#60: Add config option to disable Via header
 * BB#61: SIGHUP does not refresh filter list
 * BB#62: Make tinyproxy reload the config upon SIGHUP
 * BB#64: Config parsing error with reverse proxy option
 * BB#65: Format string compile warnings
 * BB#67: ACL processing error with multiple Allow statements

Contributors
~~~~~~~~~~~~

David Shanks, Mathew Mrosko, Michael Adam, Mukund Sivaraman.


Version 1.7.1
-------------

 * Fixed all warnings reported by GCC.
 * The tinyproxy manpage has been extended and converted to asciidoc.
 * There is a new tinyproxy.conf manpage that describes all the options.
 * The build system has been considerably cleaned up.
 * Various other bugs have been fixed.

Bugs fixed since version 1.7.0
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 * BB#2: Fix Tinyproxy for requests like www.site.com:8001
 * BB#5: Move templates from the doc directory to its own directory
 * BB#8: Update README, INSTALL, NEWS and the manpage
 * BB#10: Do not filter out transfer-encoding header
 * BB#18: Fix pointer aliasing issues
 * BB#53: Add a GPLv2 COPYING file

Contributors
~~~~~~~~~~~~

Andrew Stribblehill, Jeremy Hinegardner, Matthew Dempsky, Michael Adam,
Mukund Sivaraman, Robert James Kaes.


Version 1.7.0
-------------

 * There is now support for reverse proxying.
 * Tinyproxy does not bundle a vendor regular expressions library
   anymore. It uses the system installed regular expressions library.
 * The documentation has been updated.
 * Tinyproxy now contains some code optimizations such as the use of a
   hashmap internally for looking up error pages.
 * Various other bugs have been fixed.

Contributors
~~~~~~~~~~~~

Kim Holviala, Marc Silver, Robert James Kaes, Steven Young.


src/tinyproxy.c
/* $Id: tinyproxy.c,v 1.46 2003-03-17 04:24:19 rjkaes Exp $
 *
 * The initialize routine. Basically sets up all the initial stuff (logfile,
 * listening socket, config options, etc.) and then sits there and loops
 * over the new connections until the daemon is closed. Also has additional
 * functions to handle the "user friendly" aspects of a program (usage,
 * stats, etc.) Like any good program, most of the work is actually done
 * elsewhere.
 *
 * Copyright (C) 1998       Steven Young
 * Copyright (C) 1998-2002  Robert James Kaes (rjkaes@flarenet.com)
 * Copyright (C) 2000       Chris Lightfoot (chris@ex-parrot.com>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 */

#include "tinyproxy.h"

#include "anonymous.h"
#include "buffer.h"
#include "daemon.h"
#include "heap.h"
#include "filter.h"
#include "child.h"
#include "log.h"
#include "reqs.h"
#include "sock.h"
#include "stats.h"
#include "utils.h"

void takesig(int sig);

extern int yyparse(void);
extern FILE *yyin;

/* 
 * Global Structures
 */
struct config_s config;
float load = 0.00;
unsigned int received_sighup = FALSE; /* boolean */
unsigned int processed_config_file = FALSE; /* boolean */

/*
 * Handle a signal
 */
void
takesig(int sig)
{
	pid_t pid;
	int status;

	switch (sig) {
	case SIGHUP:
		received_sighup = TRUE;
		break;

	case SIGTERM:
		config.quit = TRUE;
		break;

	case SIGCHLD:
		while ((pid = waitpid(-1, &status, WNOHANG)) > 0)
			;
		break;
	}

	return;
}

/*
 * Display the version information for the user.
 */
static void
display_version(void)
{
	printf("%s %s (%s)\n", PACKAGE, VERSION, TARGET_SYSTEM);
}

/*
 * Display the copyright and license for this program.
 */
static void
display_license(void)
{
	display_version();

	printf("\
  Copyright 1998       Steven Young (sdyoung@well.com)\n\
  Copyright 1998-2002  Robert James Kaes (rjkaes@users.sourceforge.net)\n\
  Copyright 1999       George Talusan (gstalusan@uwaterloo.ca)\n\
  Copyright 2000       Chris Lightfoot (chris@ex-parrot.com)\n\
\n\
  This program is free software; you can redistribute it and/or modify\n\
  it under the terms of the GNU General Public License as published by\n\
  the Free Software Foundation; either version 2, or (at your option)\n\
  any later version.\n\
\n\
  This program is distributed in the hope that it will be useful,\n\
  but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
  GNU General Public License for more details.\n\
\n\
  You should have received a copy of the GNU General Public License\n\
  along with this program; if not, write to the Free Software\n\
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.\n");
}

/*
 * Display usage to the user.
 */
static void
display_usage(void)
{
	printf("Usage: %s [options]\n", PACKAGE);
	printf("\
Options:\n\
  -d		Operate in DEBUG mode.\n\
  -c FILE	Use an alternate configuration file.\n\
  -h		Display this usage information.\n\
  -l            Display the license.\n\
  -v            Display the version number.\n");

	/* Display the modes compiled into tinyproxy */
	printf("\nFeatures Compiled In:\n");
#ifdef XTINYPROXY_ENABLE
	printf("    XTinyproxy Header\n");
#endif				/* XTINYPROXY */
#ifdef FILTER_ENABLE
	printf("    Filtering\n");
#endif				/* FILTER_ENABLE */
#ifndef NDEBUG
	printf("    Debugging code\n");
#endif				/* NDEBUG */
#ifdef TRANSPARENT_PROXY
	printf("    Transparent Proxy Support\n");
#endif                          /* TRANSPARENT_PROXY */
}

int
main(int argc, char **argv)
{
	int optch;
	unsigned int godaemon = TRUE; /* boolean */
	struct passwd *thisuser = NULL;
	struct group *thisgroup = NULL;

	/*
	 * Disable the creation of CORE files right up front.
	 */
#if defined(HAVE_SETRLIMIT) && defined(NDEBUG)
	struct rlimit core_limit = { 0, 0 };
	if (setrlimit(RLIMIT_CORE, &core_limit) < 0) {
		fprintf(stderr, "%s: Could not set the core limit to zero.\n",
			argv[0]);
		exit(EX_SOFTWARE);
	}
#endif				/* HAVE_SETRLIMIT */

	/* Default configuration file location */
	config.config_file = DEFAULT_CONF_FILE;

	/*
	 * Process the various options
	 */
	while ((optch = getopt(argc, argv, "c:vldh")) != EOF) {
		switch (optch) {
		case 'v':
			display_version();
			exit(EX_OK);
		case 'l':
			display_license();
			exit(EX_OK);
		case 'd':
			godaemon = FALSE;
			break;
		case 'c':
			config.config_file = safestrdup(optarg);
			if (!config.config_file) {
				fprintf(stderr,
					"%s: Could not allocate memory.\n",
					argv[0]);
				exit(EX_SOFTWARE);
			}
			break;
		case 'h':
		default:
			display_usage();
			exit(EX_OK);
		}
	}

	log_message(LOG_INFO, "Initializing " PACKAGE " ...");

	/*
	 * Make sure the HTML error pages array is NULL to begin with.
	 * (FIXME: Should have a better API for all this)
	 */
	config.errorpages = NULL;

	/*
	 * Read in the settings from the config file.
	 */
	yyin = fopen(config.config_file, "r");
	if (!yyin) {
		fprintf(stderr,
			"%s: Could not open configuration file \"%s\".\n",
			argv[0], config.config_file);
		exit(EX_SOFTWARE);
	}
	yyparse();

	/* Open the log file if not using syslog */
	if (config.syslog == FALSE) {
		if (!config.logf_name) {
			fprintf(stderr,
				"%s: You MUST set a LogFile in the configuration file.\n",
				argv[0]);
			exit(EX_SOFTWARE);
		} else {
			if (open_log_file(config.logf_name) < 0) {
				fprintf(stderr,
					"%s: Could not create log file.\n",
					argv[0]);
				exit(EX_SOFTWARE);
			}
		}
	} else {
		if (godaemon == TRUE)
			openlog("tinyproxy", LOG_PID, LOG_DAEMON);
		else
			openlog("tinyproxy", LOG_PID, LOG_USER);
	}

	processed_config_file = TRUE;
	send_stored_logs();

	/*
	 * Set the default values if they were not set in the config file.
	 */
	if (config.port == 0) {
		fprintf(stderr,
			"%s: You MUST set a Port in the configuration file.\n",
			argv[0]);
		exit(EX_SOFTWARE);
	}
	if (!config.stathost) {
		log_message(LOG_INFO, "Setting stathost to \"%s\".",
			    DEFAULT_STATHOST);
		config.stathost = DEFAULT_STATHOST;
	}
	if (!config.username) {
		log_message(LOG_WARNING,
			    "You SHOULD set a UserName in the configuration file. Using current user instead.");
	}
	if (config.idletimeout == 0) {
		log_message(LOG_WARNING, "Invalid idle time setting. Only values greater than zero allowed; therefore setting idle timeout to %u seconds.",
			    MAX_IDLE_TIME);
		config.idletimeout = MAX_IDLE_TIME;
	}

	init_stats();

	/*
	 * If ANONYMOUS is turned on, make sure that Content-Length is
	 * in the list of allowed headers, since it is required in a
	 * HTTP/1.0 request. Also add the Content-Type header since it goes
	 * hand in hand with Content-Length.
	 *     - rjkaes
	 */
	if (is_anonymous_enabled()) {
		anonymous_insert("Content-Length");
		anonymous_insert("Content-Type");
	}

	if (godaemon == TRUE)
		makedaemon();

	if (config.pidpath) {
		if (pidfile_create(config.pidpath) < 0) {
			fprintf(stderr, "%s: Could not create PID file.\n",
				argv[0]);
			exit(EX_OSERR);
		}
	}

	if (set_signal_handler(SIGPIPE, SIG_IGN) == SIG_ERR) {
		fprintf(stderr, "%s: Could not set the \"SIGPIPE\" signal.\n",
			argv[0]);
		exit(EX_OSERR);
	}

#ifdef FILTER_ENABLE
	if (config.filter)
		filter_init();
#endif				/* FILTER_ENABLE */

	/*
	 * Start listening on the selected port.
	 */
	if (child_listening_sock(config.port) < 0) {
		fprintf(stderr, "%s: Could not create listening socket.\n",
			argv[0]);
		exit(EX_OSERR);
	}

	/*
	 * Switch to a different user.
	 */
	if (geteuid() == 0) {
		if (config.group && strlen(config.group) > 0) {
			thisgroup = getgrnam(config.group);
			if (!thisgroup) {
				fprintf(stderr,
					"%s: Unable to find group \"%s\".\n",
					argv[0], config.group);
				exit(EX_NOUSER);
			}
			if (setgid(thisgroup->gr_gid) < 0) {
				fprintf(stderr,
					"%s: Unable to change to group \"%s\".\n",
					argv[0], config.group);
				exit(EX_CANTCREAT);
			}
			log_message(LOG_INFO, "Now running as group \"%s\".",
				    config.group);
		}
		if (config.username && strlen(config.username) > 0) {
			thisuser = getpwnam(config.username);
			if (!thisuser) {
				fprintf(stderr,
					"%s: Unable to find user \"%s\".",
					argv[0], config.username);
				exit(EX_NOUSER);
			}
			if (setuid(thisuser->pw_uid) < 0) {
				fprintf(stderr,
					"%s: Unable to change to user \"%s\".",
					argv[0], config.username);
				exit(EX_CANTCREAT);
			}
			log_message(LOG_INFO, "Now running as user \"%s\".",
				    config.username);
		}
	} else {
		log_message(LOG_WARNING,
			    "Not running as root, so not changing UID/GID.");
	}

	if (child_pool_create() < 0) {
		fprintf(stderr, "%s: Could not create the pool of children.",
			argv[0]);
		exit(EX_SOFTWARE);
	}

	/*
	 * These signals are only for the parent process.
	 */
	log_message(LOG_INFO, "Setting the various signals.");
	if (set_signal_handler(SIGCHLD, takesig) == SIG_ERR) {
		fprintf(stderr, "%s: Could not set the \"SIGCHLD\" signal.\n",
			argv[0]);
		exit(EX_OSERR);
	}
	if (set_signal_handler(SIGTERM, takesig) == SIG_ERR) {
		fprintf(stderr, "%s: Could not set the \"SIGTERM\" signal.\n",
			argv[0]);
		exit(EX_OSERR);
	}
	if (set_signal_handler(SIGHUP, takesig) == SIG_ERR) {
		fprintf(stderr, "%s: Could not set the \"SIGHUP\" signal.\n",
			argv[0]);
		exit(EX_OSERR);
	}

	/*
	 * Start the main loop.
	 */
	log_message(LOG_INFO, "Starting main loop. Accepting connections.");

	child_main_loop();

	log_message(LOG_INFO, "Shutting down.");

	child_kill_children();
	child_close_sock();

	/*
	 * Remove the PID file.
	 */
	if (unlink(config.pidpath) < 0) {
		log_message(LOG_WARNING,
			    "Could not remove PID file \"%s\": %s.",
			    config.pidpath, strerror(errno));
	}

#ifdef FILTER_ENABLE
	if (config.filter)
		filter_destroy();
#endif				/* FILTER_ENABLE */

	if (config.syslog)
		closelog();
	else
		close_log_file();

	exit(EX_OK);
}
